File: misc/AECTech2024/05_Libraries/PaintHeights_Lib.py
#! python3
# https://www.nuget.org/packages/EasingFunctions
# https://github.com/emmauss/easing-functions
#r "nuget: EasingFunctions, 1.0.1"
import sys
import System
from System.Drawing import Color

import Rhino
from Rhino.Input import RhinoGet
from Rhino.Input.Custom import GetPoint, OptionDouble
from Rhino.DocObjects import ObjectType
from Rhino.Commands import Result
from Rhino.Geometry import Point3d, Ray3d, Circle, Plane, Transform, Line, MeshType

from Eto.Forms import Keyboard, Keys

from Easing import Quartic, Vector

from BrushTools import BaseBrush

class Brush(BaseBrush):
    def __init__(self, brush_surface):
        super().__init__()

        self._bsurface = brush_surface

    def GetPaintSurfaces(self):
        return self._bsurface.GetMeshes(MeshType.Render)

    def ApplyBrush(self, doc, at, reverse):
        for robj in doc.Objects.FindByObjectType(ObjectType.Extrusion):
            bbox = robj.Geometry.GetBoundingBox(True)
            pp = bbox.Min
            dist = at.DistanceTo(pp)
            if dist < self._falloff:
                height = bbox.Max.Z - bbox.Min.Z
                value = self.ComputeValue(dist, height, reverse)
                xform = Transform.Scale(Plane.WorldXY, 1.0, 1.0, value / height)
                doc.Objects.Transform(robj, xform, True)


success, obj_ref = RhinoGet.GetOneObject("Select Brush Plane", False, filter=ObjectType.Brep | ObjectType.Mesh)
if success == Result.Success:
    brush_surface = obj_ref.Object()
    brush = Brush(brush_surface)
    brush.DoPaint()

File: misc/AECTech2024/05_Libraries/PaintHeights_Lib.cs
// #! csharp
using System;
using System.Linq;
using System.Collections.Generic;

using Rhino;
using Rhino.Input;
using Rhino.Input.Custom;
using Rhino.Geometry;
using Rhino.DocObjects;
using Rhino.Commands;

class Brush : BrushTools.BaseBrush
{ 
    readonly RhinoObject _bplane;

    public Brush(RhinoObject brushPlane)
    {
        _bplane = brushPlane;
    }

    protected override IEnumerable<Mesh> GetPaintSurfaces() => _bplane.GetMeshes(MeshType.Render);

    protected override void ApplyBrush(RhinoDoc doc, Point3d at, bool reverse)
    {
        foreach (RhinoObject robj in doc.Objects
                                        .FindByObjectType(ObjectType.Extrusion))
        {
            var bbox = robj.Geometry.GetBoundingBox(true);
            var pp = bbox.Min;
            double dist = at.DistanceTo(pp);
            if (dist < m_falloff)
            {
                double height = (bbox.Max - bbox.Min).Z;
                double value = ComputeValue(dist, height, reverse);

                var xform = Transform.Scale(Plane.WorldXY, 1.0d, 1.0d, value / height);
                doc.Objects.Transform(robj, xform, true);
            }
        }
    }
}


var s = RhinoGet.GetOneObject("Select Brush Plane", false, ObjectType.Brep | ObjectType.Mesh, out ObjRef objRef);
if (s == Result.Success)
{
    RhinoObject brushPlane = objRef.Object();
    var brush = new Brush(brushPlane);
    brush.DoPaint();
}
